File: argon-ast/src/main/scala/dev/argon/ast/Ast.scala
package dev.argon.ast

import cats.data.NonEmptySeq
import dev.argon.util.{WithSource, FilePosition, Location}

sealed trait Stmt
sealed trait RecordBodyStmt
sealed trait EnumBodyStmt

final case class FunctionDeclarationStmt(
  modifiers: Seq[WithSource[Modifier]],
  purity: Boolean,
  name: WithSource[Option[IdentifierExpr]],
  parameters: Seq[WithSource[FunctionParameterList]],
  returnType: WithSource[ReturnTypeSpecifier],
  body: FunctionBody,
) extends Stmt with RecordBodyStmt with EnumBodyStmt

final case class VariableDeclarationStmt(
  modifiers: Seq[WithSource[Modifier]],
  isMutable: Boolean,
  name: Option[IdentifierExpr],
  varType: Option[WithSource[Expr]],
  value: WithSource[Expr],
) extends Stmt

final case class RecordDeclarationStmt(
  modifiers: Seq[WithSource[Modifier]],
  name: WithSource[IdentifierExpr],
  parameters: Seq[WithSource[FunctionParameterList]],
  returnType: Option[WithSource[Expr]],
  body: Seq[WithSource[RecordBodyStmt]],
) extends Stmt

final case class RecordField(
  isMutable: Boolean,
  name: WithSource[IdentifierExpr],
  fieldType: WithSource[Expr],
) extends RecordBodyStmt

final case class EnumDeclarationStmt(
  modifiers: Seq[WithSource[Modifier]],
  name: WithSource[IdentifierExpr],
  parameters: Seq[WithSource[FunctionParameterList]],
  returnType: Option[WithSource[Expr]],
  body: Seq[WithSource[EnumBodyStmt]],
) extends Stmt

enum EnumVariant extends EnumBodyStmt {
  case Constructor(
    modifiers: Seq[WithSource[Modifier]],
    name: WithSource[IdentifierExpr],
    parameters: Seq[WithSource[FunctionParameterList]],
    returnType: Option[WithSource[Expr]],
  )
  case Record(record: RecordDeclarationStmt)
}

final case class MethodDeclarationStmt(
  modifiers: Seq[WithSource[Modifier]],
  purity: Boolean,
  instanceName: WithSource[Option[IdentifierExpr]],
  instanceType: Option[WithSource[Expr]],
  name: WithSource[Option[IdentifierExpr]],
  parameters: Seq[WithSource[FunctionParameterList]],
  returnType: WithSource[ReturnTypeSpecifier],
  body: Option[FunctionBody],
) extends Stmt with RecordBodyStmt with EnumBodyStmt

final case class FunctionParameter
(paramType: WithSource[Expr], name: IdentifierExpr)

enum FunctionParameterListType derives CanEqual {
  case NormalList
  case InferrableList
  case QuoteList
  case RequiresList
}

final case class FunctionParameterList
(
  listType: FunctionParameterListType,
  isErased: Boolean,
  parameters: Seq[WithSource[FunctionParameter]],
  hasTrailingComma: Boolean,
)

final case class ReturnTypeSpecifier
(
  returnType: WithSource[Expr],
  ensuresClauses: Seq[WithSource[Expr]],
)

enum Modifier derives CanEqual {
  case Public, Protected, Private, Internal
  case Erased, Witness, Inline
}

enum ImportStmt extends Stmt {
  case Absolute(path: ImportPathSegment)
  case Relative(upCount: Int, path: ImportPathSegment)
  case Tube(tubeName: NonEmptySeq[String], path: ImportPathSegment)
  case Member(memberPath: ImportPathSegment)
}

sealed trait ImportPathSegment derives CanEqual

object ImportPathSegment {
  sealed trait End extends ImportPathSegment derives CanEqual

  final case class Cons(id: String, subPath: ImportPathSegment) extends ImportPathSegment
  final case class Many(segments: Seq[ImportPathSegment]) extends ImportPathSegment
  final case class Renaming(importing: WithSource[IdentifierExpr], viewedName: WithSource[Option[IdentifierExpr]]) extends End
  final case class Imported(id: WithSource[IdentifierExpr]) extends End
  final case class Wildcard(location: Location[FilePosition]) extends End
}

final case class ExportStmt(fromImport: WithSource[ImportStmt]) extends Stmt


final case class AssertStmt(t: WithSource[Expr]) extends Stmt


sealed trait Expr extends Stmt

object Expr:
  final case class As(value: WithSource[Expr], valueType: WithSource[Expr]) extends Expr
  final case class Assert(t: WithSource[Expr]) extends Expr
  final case class BinaryOperation(a: WithSource[Expr], op: BinaryOperator, b: WithSource[Expr]) extends Expr
  final case class Block(body: WithSource[Seq[WithSource[Stmt]]], finallyBody: Option[WithSource[Seq[WithSource[Stmt]]]]) extends Expr
  final case class BoolLiteral(b: Boolean) extends Expr
  final case class Builtin(name: String) extends Expr
  final case class Dot(o: WithSource[Expr], member: WithSource[IdentifierExpr]) extends Expr
  final case class FunctionLiteral(parameterName: Option[IdentifierExpr], body: WithSource[Expr]) extends Expr
  final case class FunctionCall(func: WithSource[Expr], listType: FunctionParameterListType, arg: WithSource[Expr]) extends Expr
  case object FunctionResultValue extends Expr
  final case class FunctionType(a: WithSource[Expr], r: WithSource[Expr]) extends Expr
  final case class IfElse(condition: WithSource[Expr], whenTrue: WithSource[Seq[WithSource[Stmt]]], whenFalse: WithSource[Seq[WithSource[Stmt]]]) extends Expr
  final case class IntLiteral(i: BigInt) extends Expr
  final case class Is(value: WithSource[Expr], pattern: WithSource[Pattern]) extends Expr
  final case class Match(value: WithSource[Expr], cases: Seq[WithSource[MatchCase]]) extends Expr
  final case class RecordLiteral(recordExpr: WithSource[Expr], fields: WithSource[Seq[WithSource[RecordFieldLiteral]]]) extends Expr
  final case class StringLiteral(parts: Seq[StringFragment]) extends Expr
  final case class Summon(t: WithSource[Expr]) extends Expr
  final case class Tuple(items: Seq[WithSource[Expr]]) extends Expr
  case object Type extends Expr
  final case class BigType(n: BigInt) extends Expr
  final case class UnaryOperation(op: UnaryOperator, a: WithSource[Expr]) extends Expr
  final case class BoxedType(t: WithSource[Expr]) extends Expr
  final case class Box(value: WithSource[Expr]) extends Expr
  final case class Unbox(value: WithSource[Expr]) extends Expr
end Expr

enum IdentifierExpr extends Expr derives CanEqual {
  case Named(s: String)
  case Op(op: Operator.ValidIdentifier)
  case Extension(inner: IdentifierExpr)
  case Inverse(inner: IdentifierExpr)
  case Update(inner: IdentifierExpr)
}

enum FunctionBody derives CanEqual {
  case ExprBody(expr: WithSource[Expr])
  case ExternBody(name: WithSource[String])
}

final case class RecordFieldLiteral(
  name: WithSource[IdentifierExpr],
  value: WithSource[Expr],
)

enum StringFragment {
  case Text(s: String)
  case Interpolate(value: WithSource[Expr])
}

sealed trait Operator derives CanEqual:
  val symbol: String
end Operator

object Operator:
  sealed trait ValidIdentifier
end Operator


enum BinaryOperator(val symbol: String) extends Operator derives CanEqual:
  case Assign extends BinaryOperator(":=")
  case Plus extends BinaryOperator("+") with Operator.ValidIdentifier
  case Minus extends BinaryOperator("-") with Operator.ValidIdentifier
  case Mul extends BinaryOperator("×") with Operator.ValidIdentifier
  case Div extends BinaryOperator("÷") with Operator.ValidIdentifier
  case Equal extends BinaryOperator("=") with Operator.ValidIdentifier
  case NotEqual extends BinaryOperator("≠") with Operator.ValidIdentifier
  case LessThan extends BinaryOperator("<") with Operator.ValidIdentifier
  case LessThanEq extends BinaryOperator("≤") with Operator.ValidIdentifier
  case GreaterThan extends BinaryOperator(">") with Operator.ValidIdentifier
  case GreaterThanEq extends BinaryOperator("≥") with Operator.ValidIdentifier
  case BitOr extends BinaryOperator("|||") with Operator.ValidIdentifier
  case BitXOr extends BinaryOperator("^^^") with Operator.ValidIdentifier
  case BitAnd extends BinaryOperator("&&&") with Operator.ValidIdentifier
  case LogicalOr extends BinaryOperator("||")
  case LogicalAnd extends BinaryOperator("&&")
  case ShiftLeft extends BinaryOperator("<<") with Operator.ValidIdentifier
  case ShiftRight extends BinaryOperator(">>") with Operator.ValidIdentifier
  case Concat extends BinaryOperator("++") with Operator.ValidIdentifier
  case PropEqual extends BinaryOperator("==")
  case PropDisjunction extends BinaryOperator("\\/")
  case PropConjunction extends BinaryOperator("/\\")
end BinaryOperator

enum UnaryOperator(val symbol: String) extends Operator derives CanEqual:
  case Plus extends UnaryOperator("+") with Operator.ValidIdentifier
  case Minus extends UnaryOperator("-") with Operator.ValidIdentifier
  case BitNot extends UnaryOperator("~~~") with Operator.ValidIdentifier
  case LogicalNot extends UnaryOperator("!") with Operator.ValidIdentifier
end UnaryOperator

final case class MatchCase(
  pattern: WithSource[Pattern],
  body: WithSource[Expr],
)

enum Pattern derives CanEqual {
  case Discard
  case Tuple(elements: Seq[WithSource[Pattern]])
  case Binding(isMutable: Boolean, name: WithSource[IdentifierExpr], pattern: WithSource[Pattern])
  case Constructor(path: WithSource[PatternPath], args: Seq[WithSource[Pattern]])
  case Record(path: WithSource[PatternPath], args: Seq[WithSource[Pattern]], recordFieldPatterns: Seq[WithSource[RecordFieldPattern]])
}

enum PatternPath {
  case Member(base: WithSource[PatternPath], member: WithSource[IdentifierExpr])
  case Base(name: WithSource[IdentifierExpr])
}

final case class RecordFieldPattern(
  fieldName: WithSource[IdentifierExpr],
  pattern: WithSource[Pattern],
)




final case class TubeSpec(
  modules: Seq[ModulePatternMapping],
)


final case class ModulePatternMapping(
  module: Seq[WithSource[ModulePatternSegment]],
  fileNameTemplate: Expr.StringLiteral,
)


enum ModulePatternSegment {
  case Named(name: String)
  case Star(boundName: IdentifierExpr)
  case DoubleStar(boundName: IdentifierExpr)
}

final case class ModuleDeclaration(
  modulePath: Seq[String],
  stmts: Seq[WithSource[Stmt]],
)

File: argon-parser/src/ast.rs
use crate::Location;
use alloc::{boxed::Box, string::String, vec::Vec};
use core::fmt::{Display, Formatter};
use num_bigint::{BigInt, BigUint};
use parse18_runtime::WithLocation;

#[derive(Debug, Clone)]
pub struct NonEmptyVec<T> {
    pub head: T,
    pub tail: Vec<T>,
}

impl<T> NonEmptyVec<T> {
    pub fn new(head: T, tail: Vec<T>) -> Self {
        Self { head, tail }
    }
}

#[derive(Debug, Clone)]
pub enum Stmt {
    Error,
    FunctionDeclaration(Box<FunctionDeclarationStmt>),
    VariableDeclaration(Box<VariableDeclarationStmt>),
    RecordDeclaration(Box<RecordDeclarationStmt>),
    EnumDeclaration(Box<EnumDeclarationStmt>),
    TraitDeclaration(Box<TraitDeclarationStmt>),
    MethodDeclaration(Box<MethodDeclarationStmt>),
    InstanceDeclaration(Box<InstanceDeclarationStmt>),
    Import(Box<ImportStmt>),
    Export(Box<ExportStmt>),
    Assert(Box<AssertStmt>),
    Expr(WithLocation<Expr>),
}

#[derive(Debug, Clone)]
pub enum RecordBodyStmt {
    FunctionDeclaration(Box<FunctionDeclarationStmt>),
    RecordField(Box<RecordField>),
    MethodDeclaration(Box<MethodDeclarationStmt>),
}

#[derive(Debug, Clone)]
pub enum EnumBodyStmt {
    FunctionDeclaration(Box<FunctionDeclarationStmt>),
    MethodDeclaration(Box<MethodDeclarationStmt>),
    EnumVariant(Box<EnumVariant>),
}

#[derive(Debug, Clone)]
pub enum TraitBodyStmt {
    FunctionDeclaration(Box<FunctionDeclarationStmt>),
    MethodDeclaration(Box<MethodDeclarationStmt>),
}

#[derive(Debug, Clone)]
pub enum NewTraitObjectBodyStmt {
    FunctionDeclaration(Box<FunctionDeclarationStmt>),
    MethodDeclaration(Box<MethodDeclarationStmt>),
}

#[derive(Debug, Clone)]
pub enum DeclarationStmt {
    Function(Box<FunctionDeclarationStmt>),
    Record(Box<RecordDeclarationStmt>),
    Enum(Box<EnumDeclarationStmt>),
    Trait(Box<TraitDeclarationStmt>),
    Method(Box<MethodDeclarationStmt>),
    Instance(Box<InstanceDeclarationStmt>),
}

impl DeclarationStmt {
    pub fn modifiers(&self) -> &[WithLocation<Modifier>] {
        match self {
            DeclarationStmt::Function(x) => &x.modifiers,
            DeclarationStmt::Record(x) => &x.modifiers,
            DeclarationStmt::Enum(x) => &x.modifiers,
            DeclarationStmt::Trait(x) => &x.modifiers,
            DeclarationStmt::Method(x) => &x.modifiers,
            DeclarationStmt::Instance(x) => &x.modifiers,
        }
    }

    pub fn name(&self) -> &WithLocation<Identifier> {
        match self {
            DeclarationStmt::Function(x) => &x.name,
            DeclarationStmt::Record(x) => &x.name,
            DeclarationStmt::Enum(x) => &x.name,
            DeclarationStmt::Trait(x) => &x.name,
            DeclarationStmt::Method(x) => &x.name,
            DeclarationStmt::Instance(x) => &x.name,
        }
    }
}

#[derive(Debug, Clone)]
pub struct FunctionDeclarationStmt {
    pub modifiers: Vec<WithLocation<Modifier>>,
    pub purity: bool,
    pub name: WithLocation<Identifier>,
    pub parameters: Vec<WithLocation<FunctionParameterList>>,
    pub return_type: WithLocation<ReturnTypeSpecifier>,
    pub body: FunctionBody,
}

#[derive(Debug, Clone)]
pub struct VariableDeclarationStmt {
    pub modifiers: Vec<WithLocation<Modifier>>,
    pub is_mutable: bool,
    pub name: Option<Identifier>,
    pub var_type: Option<WithLocation<Expr>>,
    pub value: WithLocation<Expr>,
}

#[derive(Debug, Clone)]
pub struct RecordDeclarationStmt {
    pub modifiers: Vec<WithLocation<Modifier>>,
    pub name: WithLocation<Identifier>,
    pub parameters: Vec<WithLocation<FunctionParameterList>>,
    pub return_type: Option<WithLocation<Expr>>,
    pub body: Vec<WithLocation<RecordBodyStmt>>,
}

#[derive(Debug, Clone)]
pub struct RecordField {
    pub is_mutable: bool,
    pub name: WithLocation<Identifier>,
    pub field_type: WithLocation<Expr>,
}

#[derive(Debug, Clone)]
pub struct EnumDeclarationStmt {
    pub modifiers: Vec<WithLocation<Modifier>>,
    pub name: WithLocation<Identifier>,
    pub parameters: Vec<WithLocation<FunctionParameterList>>,
    pub return_type: Option<WithLocation<Expr>>,
    pub body: Vec<WithLocation<EnumBodyStmt>>,
}

#[derive(Debug, Clone)]
pub enum EnumVariant {
    Constructor {
        modifiers: Vec<WithLocation<Modifier>>,
        name: WithLocation<Identifier>,
        parameters: Vec<WithLocation<FunctionParameterList>>,
        return_type: Option<WithLocation<Expr>>,
    },
    Record(RecordDeclarationStmt),
}

#[derive(Debug, Clone)]
pub struct TraitDeclarationStmt {
    pub modifiers: Vec<WithLocation<Modifier>>,
    pub name: WithLocation<Identifier>,
    pub parameters: Vec<WithLocation<FunctionParameterList>>,
    pub return_type: Option<WithLocation<Expr>>,
    pub body: Vec<WithLocation<TraitBodyStmt>>,
}

#[derive(Debug, Clone)]
pub struct MethodDeclarationStmt {
    pub modifiers: Vec<WithLocation<Modifier>>,
    pub purity: bool,
    pub instance_name: WithLocation<Option<Identifier>>,
    pub instance_type: Option<WithLocation<Expr>>,
    pub name: WithLocation<Identifier>,
    pub parameters: Vec<WithLocation<FunctionParameterList>>,
    pub return_type: WithLocation<ReturnTypeSpecifier>,
    pub body: Option<FunctionBody>,
}

#[derive(Debug, Clone)]
pub struct InstanceDeclarationStmt {
    pub modifiers: Vec<WithLocation<Modifier>>,
    pub name: WithLocation<Identifier>,
    pub parameters: Vec<WithLocation<FunctionParameterList>>,
    pub return_type: Option<WithLocation<Expr>>,
    pub body: Vec<WithLocation<TraitBodyStmt>>,
}

#[derive(Debug, Clone)]
pub struct FunctionParameter {
    pub param_type: WithLocation<Expr>,
    pub name: Identifier,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FunctionParameterListType {
    NormalList,
    InferrableList(BigUint),
    RequiresList,
}

#[derive(Debug, Clone)]
pub struct FunctionParameterList {
    pub list_type: FunctionParameterListType,
    pub modifiers: Vec<WithLocation<Modifier>>,
    pub parameters: Vec<WithLocation<FunctionParameter>>,
    pub has_trailing_comma: bool,
}

#[derive(Debug, Clone)]
pub struct ReturnTypeSpecifier {
    pub return_type: WithLocation<Expr>,
    pub ensures_clauses: Vec<WithLocation<Expr>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Modifier {
    Public,
    Internal,
    Protected,
    Private,
    Erased,
    Token,
    Witness,
    Inline,
    Final,
    Override,
    Virtual,
}

impl Display for Modifier {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Modifier::Public => write!(f, "public"),
            Modifier::Internal => write!(f, "internal"),
            Modifier::Protected => write!(f, "protected"),
            Modifier::Private => write!(f, "private"),
            Modifier::Erased => write!(f, "erased"),
            Modifier::Token => write!(f, "token"),
            Modifier::Witness => write!(f, "witness"),
            Modifier::Inline => write!(f, "inline"),
            Modifier::Final => write!(f, "final"),
            Modifier::Override => write!(f, "override"),
            Modifier::Virtual => write!(f, "virtual"),
        }
    }
}

#[derive(Debug, Clone)]
pub enum ImportStmt {
    Absolute(ImportPathSegment),
    Relative {
        up_count: usize,
        path: ImportPathSegment,
    },
    Tube {
        tube_name: NonEmptyVec<String>,
        path: ImportPathSegment,
    },
    Member {
        member_path: ImportPathSegment,
    },
}

#[derive(Debug, Clone)]
pub enum ImportPathSegment {
    Cons {
        id: String,
        sub_path: Box<ImportPathSegment>,
    },
    Many {
        segments: Vec<ImportPathSegment>,
    },
    Renaming {
        importing: WithLocation<Identifier>,
        viewed_name: WithLocation<Option<Identifier>>,
    },
    Imported {
        id: WithLocation<Identifier>,
    },
    Wildcard {
        location: Location,
    },
}

#[derive(Debug, Clone)]
pub struct ExportStmt {
    pub from_import: WithLocation<ImportStmt>,
}

#[derive(Debug, Clone)]
pub struct AssertStmt {
    pub t: WithLocation<Expr>,
}

#[derive(Debug, Clone)]
pub enum Expr {
    Error,
    As {
        value: Box<WithLocation<Expr>>,
        value_type: Box<WithLocation<Expr>>,
    },
    Assert {
        t: Box<WithLocation<Expr>>,
    },
    BinaryOperation {
        a: Box<WithLocation<Expr>>,
        op: WithLocation<BinaryOperator>,
        b: Box<WithLocation<Expr>>,
    },
    Block {
        body: WithLocation<Vec<WithLocation<Stmt>>>,
        finally_body: Option<WithLocation<Vec<WithLocation<Stmt>>>>,
    },
    BoolLiteral(bool),
    Break {
        label: Option<WithLocation<Identifier>>,
        value: Option<Box<WithLocation<Expr>>>,
    },
    Builtin(String),
    Dot {
        o: Box<WithLocation<Expr>>,
        member: WithLocation<Identifier>,
    },
    FunctionLiteral(FunctionLiteral),
    FunctionCall {
        func: Box<WithLocation<Expr>>,
        list_type: FunctionParameterListType,
        arg: Box<WithLocation<Expr>>,
    },
    FunctionResultValue,
    FunctionType {
        a: Box<WithLocation<Expr>>,
        r: Box<WithLocation<Expr>>,
    },
    IfElse {
        condition: Box<WithLocation<Expr>>,
        when_true: WithLocation<Vec<WithLocation<Stmt>>>,
        when_false: WithLocation<Vec<WithLocation<Stmt>>>,
    },
    IntLiteral(BigInt),
    Is {
        value: Box<WithLocation<Expr>>,
        pattern: WithLocation<Pattern>,
    },
    Loop {
        label: Option<WithLocation<Identifier>>,
        body: WithLocation<Vec<WithLocation<Stmt>>>,
    },
    Match {
        value: Box<WithLocation<Expr>>,
        cases: Vec<WithLocation<MatchCase>>,
    },
    NewTraitObject {
        trait_expr: Box<WithLocation<Expr>>,
        body: Vec<WithLocation<NewTraitObjectBodyStmt>>,
    },
    Next {
        label: Option<WithLocation<Identifier>>,
    },
    Paren(Box<WithLocation<Expr>>),
    Raise {
        ex: Box<WithLocation<Expr>>,
    },
    RecordLiteral {
        record_expr: Box<WithLocation<Expr>>,
        fields: WithLocation<Vec<WithLocation<RecordFieldLiteral>>>,
    },
    Redo {
        label: Option<WithLocation<Identifier>>,
    },
    Retry {
        label: Option<WithLocation<Identifier>>,
    },
    StringLiteral(StringLiteral),
    Summon {
        t: Box<WithLocation<Expr>>,
    },
    Tuple {
        items: Vec<WithLocation<Expr>>,
    },
    Type,
    BigType(BigInt),
    UnaryOperation {
        op: WithLocation<UnaryOperator>,
        a: Box<WithLocation<Expr>>,
    },
    While {
        label: Option<WithLocation<Identifier>>,
        condition: WithLocation<Vec<WithLocation<Stmt>>>,
        body: WithLocation<Vec<WithLocation<Stmt>>>,
    },
    BoxedType {
        t: Box<WithLocation<Expr>>,
    },
    Box {
        value: Box<WithLocation<Expr>>,
    },
    Unbox {
        value: Box<WithLocation<Expr>>,
    },
    Identifier(Identifier),
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Identifier {
    Named(String),
    BinaryOp(BinaryOperatorIdentifier),
    UnaryOp(UnaryOperatorIdentifier),
    Index,
    Extension(Box<Identifier>),
    Inverse(Box<Identifier>),
    Update(Box<Identifier>),
}

impl Display for Identifier {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Identifier::Named(name) => write!(f, "{}", name),
            Identifier::BinaryOp(op) => write!(f, "binary operator {}", op.symbol()),
            Identifier::UnaryOp(op) => write!(f, "unary operator {}", op.symbol()),
            Identifier::Index => write!(f, "index"),
            Identifier::Extension(inner) => write!(f, "extension {}", inner),
            Identifier::Inverse(inner) => write!(f, "inverse {}", inner),
            Identifier::Update(inner) => write!(f, "update {}", inner),
        }
    }
}

#[derive(Debug, Clone)]
pub enum FunctionBody {
    ExprBody(Box<WithLocation<Expr>>),
    ExternBody(WithLocation<String>),
}

#[derive(Debug, Clone)]
pub struct RecordFieldLiteral {
    pub name: WithLocation<Identifier>,
    pub value: WithLocation<Expr>,
}

#[derive(Debug, Clone)]
pub struct FunctionLiteral {
    pub parameter_name: Option<Identifier>,
    pub body: Box<WithLocation<Expr>>,
}

#[derive(Debug, Clone)]
pub struct StringLiteral {
    pub parts: Vec<StringFragment>,
}

#[derive(Debug, Clone)]
pub enum StringFragment {
    Text(String),
    Interpolate { value: Box<WithLocation<Expr>> },
}

pub trait Operator: Copy {
    fn symbol(self) -> &'static str;
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum BinaryOperatorIdentifier {
    Plus,
    Minus,
    Mul,
    Div,
    Equal,
    NotEqual,
    LessThan,
    LessThanEq,
    GreaterThan,
    GreaterThanEq,
    BitOr,
    BitXOr,
    BitAnd,
    ShiftLeft,
    ShiftRight,
    Concat,
}

impl TryFrom<BinaryOperator> for BinaryOperatorIdentifier {
    type Error = ();

    fn try_from(op: BinaryOperator) -> Result<Self, Self::Error> {
        match op {
            BinaryOperator::Plus => Ok(BinaryOperatorIdentifier::Plus),
            BinaryOperator::Minus => Ok(BinaryOperatorIdentifier::Minus),
            BinaryOperator::Mul => Ok(BinaryOperatorIdentifier::Mul),
            BinaryOperator::Div => Ok(BinaryOperatorIdentifier::Div),
            BinaryOperator::Equal => Ok(BinaryOperatorIdentifier::Equal),
            BinaryOperator::NotEqual => Ok(BinaryOperatorIdentifier::NotEqual),
            BinaryOperator::LessThan => Ok(BinaryOperatorIdentifier::LessThan),
            BinaryOperator::LessThanEq => Ok(BinaryOperatorIdentifier::LessThanEq),
            BinaryOperator::GreaterThan => Ok(BinaryOperatorIdentifier::GreaterThan),
            BinaryOperator::GreaterThanEq => Ok(BinaryOperatorIdentifier::GreaterThanEq),
            BinaryOperator::BitOr => Ok(BinaryOperatorIdentifier::BitOr),
            BinaryOperator::BitXOr => Ok(BinaryOperatorIdentifier::BitXOr),
            BinaryOperator::BitAnd => Ok(BinaryOperatorIdentifier::BitAnd),
            BinaryOperator::ShiftLeft => Ok(BinaryOperatorIdentifier::ShiftLeft),
            BinaryOperator::ShiftRight => Ok(BinaryOperatorIdentifier::ShiftRight),
            BinaryOperator::Concat => Ok(BinaryOperatorIdentifier::Concat),
            BinaryOperator::Assign
            | BinaryOperator::LogicalOr
            | BinaryOperator::LogicalAnd
            | BinaryOperator::PropEqual
            | BinaryOperator::PropDisjunction
            | BinaryOperator::PropConjunction => Err(()),
        }
    }
}

impl From<BinaryOperatorIdentifier> for BinaryOperator {
    fn from(op: BinaryOperatorIdentifier) -> Self {
        match op {
            BinaryOperatorIdentifier::Plus => BinaryOperator::Plus,
            BinaryOperatorIdentifier::Minus => BinaryOperator::Minus,
            BinaryOperatorIdentifier::Mul => BinaryOperator::Mul,
            BinaryOperatorIdentifier::Div => BinaryOperator::Div,
            BinaryOperatorIdentifier::Equal => BinaryOperator::Equal,
            BinaryOperatorIdentifier::NotEqual => BinaryOperator::NotEqual,
            BinaryOperatorIdentifier::LessThan => BinaryOperator::LessThan,
            BinaryOperatorIdentifier::LessThanEq => BinaryOperator::LessThanEq,
            BinaryOperatorIdentifier::GreaterThan => BinaryOperator::GreaterThan,
            BinaryOperatorIdentifier::GreaterThanEq => BinaryOperator::GreaterThanEq,
            BinaryOperatorIdentifier::BitOr => BinaryOperator::BitOr,
            BinaryOperatorIdentifier::BitXOr => BinaryOperator::BitXOr,
            BinaryOperatorIdentifier::BitAnd => BinaryOperator::BitAnd,
            BinaryOperatorIdentifier::ShiftLeft => BinaryOperator::ShiftLeft,
            BinaryOperatorIdentifier::ShiftRight => BinaryOperator::ShiftRight,
            BinaryOperatorIdentifier::Concat => BinaryOperator::Concat,
        }
    }
}

impl Operator for BinaryOperatorIdentifier {
    fn symbol(self) -> &'static str {
        BinaryOperator::from(self).symbol()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum UnaryOperatorIdentifier {
    Plus,
    Minus,
    BitNot,
}

impl TryFrom<UnaryOperator> for UnaryOperatorIdentifier {
    type Error = ();

    fn try_from(value: UnaryOperator) -> Result<Self, Self::Error> {
        Ok(match value {
            UnaryOperator::Plus => UnaryOperatorIdentifier::Plus,
            UnaryOperator::Minus => UnaryOperatorIdentifier::Minus,
            UnaryOperator::BitNot => UnaryOperatorIdentifier::BitNot,
            UnaryOperator::LogicalNot => return Err(()),
        })
    }
}

impl From<UnaryOperatorIdentifier> for UnaryOperator {
    fn from(op: UnaryOperatorIdentifier) -> Self {
        match op {
            UnaryOperatorIdentifier::Plus => UnaryOperator::Plus,
            UnaryOperatorIdentifier::Minus => UnaryOperator::Minus,
            UnaryOperatorIdentifier::BitNot => UnaryOperator::BitNot,
        }
    }
}

impl Operator for UnaryOperatorIdentifier {
    fn symbol(self) -> &'static str {
        UnaryOperator::from(self).symbol()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum BinaryOperator {
    Assign,
    Plus,
    Minus,
    Mul,
    Div,
    Equal,
    NotEqual,
    LessThan,
    LessThanEq,
    GreaterThan,
    GreaterThanEq,
    BitOr,
    BitXOr,
    BitAnd,
    LogicalOr,
    LogicalAnd,
    ShiftLeft,
    ShiftRight,
    Concat,
    PropEqual,
    PropDisjunction,
    PropConjunction,
}

impl BinaryOperator {
    #[allow(clippy::match_same_arms, reason = "Readability")]
    pub fn is_valid_identifier(self) -> bool {
        match self {
            BinaryOperator::Assign
            | BinaryOperator::LogicalOr
            | BinaryOperator::LogicalAnd
            | BinaryOperator::PropEqual
            | BinaryOperator::PropDisjunction
            | BinaryOperator::PropConjunction => false,

            BinaryOperator::Plus
            | BinaryOperator::Minus
            | BinaryOperator::Mul
            | BinaryOperator::Div
            | BinaryOperator::Equal
            | BinaryOperator::NotEqual
            | BinaryOperator::LessThan
            | BinaryOperator::LessThanEq
            | BinaryOperator::GreaterThan
            | BinaryOperator::GreaterThanEq
            | BinaryOperator::BitOr
            | BinaryOperator::BitXOr
            | BinaryOperator::BitAnd
            | BinaryOperator::ShiftLeft
            | BinaryOperator::ShiftRight
            | BinaryOperator::Concat => true,
        }
    }
}

impl Operator for BinaryOperator {
    fn symbol(self) -> &'static str {
        match self {
            BinaryOperator::Assign => ":=",
            BinaryOperator::Plus => "+",
            BinaryOperator::Minus => "-",
            BinaryOperator::Mul => "×",
            BinaryOperator::Div => "÷",
            BinaryOperator::Equal => "=",
            BinaryOperator::NotEqual => "≠",
            BinaryOperator::LessThan => "<",
            BinaryOperator::LessThanEq => "≤",
            BinaryOperator::GreaterThan => ">",
            BinaryOperator::GreaterThanEq => "≥",
            BinaryOperator::BitOr => "|||",
            BinaryOperator::BitXOr => "^^^",
            BinaryOperator::BitAnd => "&&&",
            BinaryOperator::LogicalOr => "||",
            BinaryOperator::LogicalAnd => "&&",
            BinaryOperator::ShiftLeft => "<<",
            BinaryOperator::ShiftRight => ">>",
            BinaryOperator::Concat => "++",
            BinaryOperator::PropEqual => "==",
            BinaryOperator::PropDisjunction => "\\/",
            BinaryOperator::PropConjunction => "/\\",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum UnaryOperator {
    Plus,
    Minus,
    BitNot,
    LogicalNot,
}

impl UnaryOperator {
    pub fn is_valid_identifier(self) -> bool {
        self != UnaryOperator::LogicalNot
    }
}

impl Operator for UnaryOperator {
    fn symbol(self) -> &'static str {
        match self {
            UnaryOperator::Plus => "+",
            UnaryOperator::Minus => "-",
            UnaryOperator::BitNot => "~~~",
            UnaryOperator::LogicalNot => "!",
        }
    }
}

#[derive(Debug, Clone)]
pub struct MatchCase {
    pub pattern: WithLocation<Pattern>,
    pub body: WithLocation<Expr>,
}

#[derive(Debug, Clone)]
pub enum Pattern {
    Discard,
    Tuple {
        elements: Vec<WithLocation<Pattern>>,
    },
    Binding {
        is_mutable: bool,
        name: WithLocation<Identifier>,
        pattern: Box<WithLocation<Pattern>>,
    },
    Constructor {
        path: WithLocation<PatternPath>,
        args: Vec<PatternArgument>,
    },
    Record {
        path: WithLocation<PatternPath>,
        args: Vec<PatternArgument>,
        record_field_patterns: Vec<WithLocation<RecordFieldPattern>>,
    },
    String(StringLiteral),
    Int(BigInt),
    Bool(bool),
}

#[derive(Debug, Clone)]
pub enum PatternPath {
    Member {
        base: Box<WithLocation<PatternPath>>,
        member: WithLocation<Identifier>,
    },
    Base {
        name: WithLocation<Identifier>,
    },
}

#[derive(Debug, Clone)]
pub struct RecordFieldPattern {
    pub field_name: WithLocation<Identifier>,
    pub pattern: WithLocation<Pattern>,
}

#[derive(Debug, Clone)]
pub struct PatternArgument {
    pub function_parameter_list_type: FunctionParameterListType,
    pub arg: WithLocation<Pattern>,
}

#[derive(Debug, Clone)]
pub struct TubeSpec {
    pub modules: Vec<ModulePatternMapping>,
}

#[derive(Debug, Clone)]
pub struct ModulePatternMapping {
    pub module: Vec<WithLocation<ModulePatternSegment>>,
    pub file_name_template: StringLiteral,
}

#[derive(Debug, Clone)]
pub enum ModulePatternSegment {
    Named { name: String },
    Star { bound_name: Identifier },
    DoubleStar { bound_name: Identifier },
}

#[derive(Debug, Clone)]
pub struct ModuleDeclaration {
    pub module_path: Vec<String>,
    pub stmts: Vec<WithLocation<Stmt>>,
}
